Enhance game items

charmyt opened the pull request:
Automatically scale item images so item author does not need to specify size and scaling.
Add item name and item id.
Add player inventory. (No UI yet.)
Add ability for player to pick up items and add them to inventory.

zethon commented:
How do users pickup items? Can players drop items once they've picked them up? Is there a limit to the inventory?

zethon commented on src/Item.cpp:
```diff
@@ -4,26 +4,42 @@
         return (std::string)(_json[0].at("description"));
     }
 
+    bool Item::isObtainable() { 
```
newlines!

zethon commented on src/Player.cpp:
```diff
@@ -58,4 +58,25 @@ void Player::setGlobalBottom(float bottom)
-} // namespace tt
\ No newline at end of file
+
+void Player::addItem(std::string itemID) {
```
Also, I just noticed you're passing the `std::string` by value:

```
void Player::addItem(std::string itemID) 
```

which you probably don't want since it creates an unnecessary copy of `std::string`. Instead use a const-reference:

```
void  Player::addItem(const std::string& itemID)
```

This avoids the extra copy.

zethon commented on src/Item.cpp:
```diff
@@ -4,26 +4,42 @@
 {
 
-    Item::Item( sf::Texture texture,
+    Item::Item( std::string id,
```
It is generally not advisable to pass non-trivial objects by value, as it creates a copy, and in most cases you don't need a copy. 

Instead use a `const-ref` like so:

```
Item::Item(const std::string& id, ...
```

This avoids an extra copy.

zethon commented on src/Item.cpp:
```diff
@@ -4,26 +4,42 @@
+        if(_json[0]["obtainable"].is_null()) {
+            return false;
+        }
+        return (bool)(_json[0].at("obtainable"));
```
Avoid C-style casts. Once I turn on strict compilation, this kind of thing will generate an error (really, it will generate a warning, but I will turn on "treat warnings as errors", too). Instead, C++ wants you to be explicit in the type of cast you want to do. In this case you could do:

```
return static_cast<bool>(_json[0].at("obtainable"));
```

Though, doesn't this JSON library allow you to do something like:

```
return _json[0].at("obtainable").as<bool>();
```

Or something like that...

zethon commented:
Sorry if it seems like I'm being picky with my code reviews. I'm just reviewing with the same scrutiny that I'd review someone's code in my day job.

zethon commented on src/Item.cpp:
```diff
@@ -4,26 +4,42 @@
 
-    Item::Item( sf::Texture texture,
+    Item::Item( std::string id,
+                sf::Texture texture,
```
What I mentioned above about passing by-value, applies here too. This constructor should look like:

```
Item::Item(const std::string& id, const sf::Texture& texture...
```

You get the point.

charmyt replied:
`Vehicle(sf::Texture texture,`

zethon replied:
Yup, looks like I forgot one too.

zethon commented on src/Item.cpp:
```diff
@@ -4,26 +4,42 @@
-        // first item?
+        // and we must use _json[0] to get the first item?
         //
+        return (std::string)(_json[0].at("name"));
```
Also, see my note below about the use of the C-style case.

zethon commented:
@charmyt Sorry if it seems like I'm being picky with my code reviews. I'm just reviewing with the same scrutiny that I'd review someone's code in my day job.

charmyt commented:
I'd like to pull that branch from the fork into a branch on this repo. Can I do that? 

```
amb@comet:~/git-clones/ttvg$ git push origin enhance-game-items 
ERROR: Permission to zethon/ttvg.git denied to charmyt.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
amb@comet:~/git-clones/ttvg$ 

```

zethon commented:
Could the issue be you still haven't accepted my invite to the repo?

![image](https://user-images.githubusercontent.com/5203583/88315265-15e28d00-cce4-11ea-8d4c-40b015c75806.png)

charmyt commented:
What is the crash you are seeing? Are you doing a clean build when switching branches within the same repo clone?

zethon commented:
It was a problem with your JSON. I haven't tried it since this morning though. I'll work on the game some tonight and let you know.

zethon commented on src/Item.h:
```diff
@@ -15,10 +15,10 @@ namespace tt
 
-            Item(       std::string id,
-                        sf::Texture texture, 
+            Item(       const std::string&  id,
```
I lead you astray in this instance because I didn't realize `Items` was subclassed from `AnimatedSprite`. Passing the `Texture` reference here **may** not work or cause undefined behavior because you're trying to *move* a temporary reference. Then again, it might be fine. Also, it might depend on the compiler/platform. I'm not really sure. Best bet is probably to change *this particular one* back to how you had it.

And to really mess with you, the `nl::json` object probably **should** be a `const-ref`. 🤣

charmyt replied:
I'll fix it.

charmyt replied:
I hate C++

zethon replied:
C++ is awesome once you learn the quirks.

charmyt commented:
I have addressed the review items mentioned above. 
It's quite a list so I might have missed a few. Let me know.

Also is this not okay to use the same feature branch and continue to merge the same branch to your master multiple times? I'd rather not be creating branches locally every other day. 
Maybe I should just stay on my own repo, sync to my master and do PRs from there?

zethon commented:
> Also is this not okay to use the same feature branch and continue to merge the same branch to your master multiple times? I'd rather not be creating branches locally every other day.
Maybe I should just stay on my own repo, sync to my master and do PRs from there?

Up to you how you want to do it. Usually once a branch has been merged to master, I consider it closed and create a new branch from the new master. But whatever works for you. The only reason I was confused is because I pulled your branch down to try and create a branch in this repo for you, and saw weird merge issues. But that's not typical practice. 

I commented on one thing I saw in your new changes. I'll take a closer look at the rest of them in a bit.

zethon commented on src/Player.cpp:
```diff
@@ -58,4 +58,17 @@ void Player::setGlobalBottom(float bottom)
+}
+
+
+std::map<std::string, std::int32_t> Player::getInventory()
```
I generally wouldn't return a map like this since this will copy the entire map (and everything in it). Instead you can return a read/write reference:
```
std::map<...>& Player::getInventory()
```

Or a read-only reference
```
const std::map<..>& Player::getInventory() const
```
I would err on the side of returning the read-only version. Notice I added `const` at the beginning and the end since this is also a `get`ter function.

I have a follow-up comment to this down where you call this function.

zethon commented on src/Scenes/Opening.cpp:
```diff
@@ -163,13 +174,72 @@ void Opening::initTraffic()
+            //
+            case sf::Keyboard::I:
+            {
+                std::map<std::string, std::int32_t> inv =
```
If you make the changes above that I recommended you will need to change this, else you'll still make a copy. If you implement the read-only/const version, you should update this to:

`const std::map<...>& inv = ...`

Or simply: `const auto& inv = ...`

charmyt replied:
Question:

When I change this to use a reference, this seems to break the use of the [] operator on the map.
A workaround I guess is to use map's at() method. Is there a better way to do this?

```
/home/amb/git-clones/charmyt.ttvg/src/Scenes/Opening.cpp:242:56: error: passing ‘const std::map<std::basic_string<char>, int>’ as ‘this’ argument discards qualifiers [-fpermissive]
  242 |                     std::cout << key << ": " << inv[key] << std::endl;
      |                                                        ^

...

  490 |       operator[](const key_type& __k)
      |       ^~~~~~~~

```

zethon replied:
The `at()` method was specifically added (in C++11 I think?) to address this situation: to allow const access to the elements.

zethon commented on src/Scenes/Opening.cpp:
```diff
@@ -163,13 +174,72 @@ void Opening::initTraffic()
+                                                    _player->getInventory();
+                std::vector<std::string> keys;
+
+                for(auto it = inv.begin(); it != inv.end(); it++)
```
Just an FYI, not a change request, there's some cool syntatic sugar in modern C++ that makes this kind of thing prettier. For example:

```
for (const auto& it : _inv)
{
    keys.push_back(it->first);
}
```

charmyt commented:
> Fix null check in isObtainable().

This might have been the crash you were seeing. (Attempt to pick up an item which cannot be picked up because 'obtainable' is not set in the json.)
This means the ItemFactory 'image-attr' is_null() checking code is also wrong, and should be using this `find() == end()` approach instead. Yet it is working right now for some reason, at least on my platform. But I will fix that too.

zethon commented on src/Item.cpp:
```diff
@@ -3,27 +3,47 @@
+
+    std::string Item::getDescription() const
+    {
+        return _json[0].at("description");
```
I'm ok with the way this currently is, but in the future it would be better to save the values in member variables and return those. 

Doing so would save the two index lookups that happen in this call. Granted, they are both O(1) lookups, but zero lookups is still better than two.

zethon commented:
@charmyt Looks good. Go ahead and merge if you're ready to do so!
